hcp/common.go

```diff
@@ -0,0 +1,145 @@
+package hcp
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-packer-service/stable/2021-04-30/models"
+	sdkpacker "github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/hashicorp/packer/internal/registry"
+	"github.com/mitchellh/mapstructure"
+
+	registryimage "github.com/hashicorp/packer-plugin-sdk/packer/registry/image"
+)
+
+func populateIteration(
+	ctx context.Context,
+	handler HCPHandler,
+) error {
+	err := handler.Bucket().Initialize(ctx)
+	if err != nil {
+		return fmt.Errorf("HCP Packer Registry build initialization failed: %s", err)
+	}
+
+	err = handler.Bucket().PopulateIteration(ctx)
+	if err != nil {
+		return fmt.Errorf("HCP Packer Registry build initialization failed: %s", err)
+	}
+
+	return nil
+}
+
+func buildStart(
+	ctx context.Context,
+	handler HCPHandler,
+	buildName string,
+) error {
+	b := handler.Bucket()
+
+	if !b.IsExpectingBuildForComponent(buildName) {
+		return fmt.Errorf("already done")
+	}
+
+	err := b.UpdateBuildStatus(ctx, buildName, models.HashicorpCloudPackerBuildStatusRUNNING)
+	if err != nil {
+		return fmt.Errorf("failed to update HCP Packer registry status for %q: %s", buildName, err)
+	}
+
+	cleanupHeartbeat, err := b.HeartbeatBuild(ctx, buildName)
+	if err != nil {
+		log.Printf("[ERROR] failed to start heartbeat function")
+	}
+
+	buildDone := make(chan struct{}, 1)
+	go func() {
+		log.Printf("[TRACE] waiting for heartbeat completion")
+		select {
+		case <-ctx.Done():
+			cleanupHeartbeat()
+			err := b.UpdateBuildStatus(
+				context.Background(),
+				buildName,
+				models.HashicorpCloudPackerBuildStatusCANCELLED)
+			if err != nil {
+				log.Printf(
+					"[ERROR] failed to update HCP Packer registry status for %q: %s",
+					buildName,
+					err)
+			}
+		case <-buildDone:
+			cleanupHeartbeat()
+		}
+		log.Printf("[TRACE] done waiting for heartbeat completion")
+	}()
+
+	b.RunningBuilds[buildName] = buildDone
+
+	return nil
+}
+
+func buildDone(
+	ctx context.Context,
+	handler HCPHandler,
+	buildName string,
+	artifacts []sdkpacker.Artifact,
+	buildErr error,
+) ([]sdkpacker.Artifact, error) {
+	b := handler.Bucket()
+
+	doneCh, ok := b.RunningBuilds[buildName]
+	if !ok {
+		log.Print("[ERROR] done build does not have an entry in the heartbeat table, state will be inconsistent.")
+	} else {
+		log.Printf("[TRACE] signal stopping heartbeats")
+		// Stop heartbeating
+		doneCh <- struct{}{}
+		log.Printf("[TRACE] stopped heartbeats")
+	}
+
+	if buildErr != nil {
+		err := b.UpdateBuildStatus(ctx, buildName, models.HashicorpCloudPackerBuildStatusFAILED)
+		if err != nil {
+			log.Printf("[ERROR] failed to update build %q status to FAILED: %s", buildName, err)
+		}
+		return artifacts, fmt.Errorf("build failed, not uploading artifacts")
```

Comment:
What would cause a failure here?  

If this is a communication hipcup should we try to dump the build artifact details for later use?

Reply:
This error here is not a communication hiccup, it's that there was already an error with the build. I guess we could return `nil,nil` here, since nothing is updated, and the artifacts are not updated, my bad.